common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp

```diff
@@ -647,8 +945,175 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
 bool ap_wlan_hal_dwpal::update_vap_credentials(
     std::list<son::wireless_utils::sBssInfoConf> &bss_info_conf_list)
 {
-    //TODO Implement #346
-    return true;
+    // The bss_info_conf_list contains list of all VAPs and STAs that have to be created
+    // on the radio. If the list is empty, there should be no VAPs left.
+    // At the moment we are taking a shortcut to make this work for certification setup.
+    // We are updating the hostapd .conf files with the info from the bss_info_conf_list
+    // and sending RELOAD command to hostapd making it update the VAPs.
+    // With that approach there are following limitations:
+    // - all the VAPs are expected to be always present in the config, but have
+    //   start_disabled=1 option if not active
+    // - any configuration change causing the hostapd config files to be re-genrated
+    //   and the hostapd processes restarted wil kill the purplMesh configuration
+
+    // The hostapd config is stored here. The items order in the header and VAPs sections
+    // has to be preserved (so no std::map and have to keep it as it is).
+    std::string fname;
+    std::vector<std::string> hostapd_config_head;
+    std::map<std::string, std::vector<std::string>> hostapd_config_vaps;
+
+    // Load hostapd config for the radio
+    if (!load_hostapd_config(m_radio_info.iface_name, fname, hostapd_config_head,
+                             hostapd_config_vaps)) {
+        LOG(ERROR) << "Autoconfiguration: no hostapd config to apply configuration!";
+        return false;
+    }
+
+    // If a Multi-AP Agent receives an AP-Autoconfiguration WSC message containing one or
+    // more M2, it shall validate each M2 (based on its 1905 AL MAC address) and configure
+    // a BSS on the corresponding radio for each of the M2. If the Multi-AP Agent is currently
+    // operating a BSS with operating parameters that do not completely match any of the M2 in
+    // the received AP-Autoconfiguration WSC message, it shall tear down that BSS.
+
+    // Start by disabling all the existent VAPs
+    for (auto &it : hostapd_config_vaps) {
+        const std::string &vap_if                    = it.first;
+        std::vector<std::string> &vap_hostapd_config = it.second;
+
+        if (!hostapd_config_set_value(vap_hostapd_config, "start_disabled", "1")) {
+            LOG(ERROR) << "Autoconfiguration: cannot disable VAP " << vap_if << ", ignoring";
+        }
+    }
```

Comment:
In the code below, we skip the non-ap VAPs, which IMO is a good thing. But that means we have to skip them here as well.

So I think a better approach maybe is to move this *after* the bss_info loop, and just disable all the remaining VAPs (still skipping the non-ap VAPs).

This is a nitpick since it will anyway be reworked later. Maybe add a TODO to remember to do this correctly when it is reworked.

Reply:
It seems like this has to be done, otherwise even experiment with the wireless backhauls will not be possible till this code is gone.

Reply:
Wireless backhaul are not VAPs. It's completely separate.

A wireless backhaul (usually) does count towards the maximum number of BSSes that can be configured on a radio, but we can only handle that by reserving a BSS slot for the bSTA.